crates/services/consensus_module/poa/src/service_test.rs

```diff
@@ -397,3 +425,124 @@ fn test_signing_key() -> Secret<SecretKeyWrapper> {
     let secret_key = SecretKey::random(&mut rng);
     Secret::new(secret_key.into())
 }
+
+#[derive(Debug, PartialEq)]
+enum FakeProducedBlock {
+    Predefined(Block),
+    New(BlockHeight, Tai64),
+}
+
+struct FakeBlockProducer {
+    block_sender: tokio::sync::mpsc::Sender<FakeProducedBlock>,
+}
+
+impl FakeBlockProducer {
+    fn new() -> (Self, tokio::sync::mpsc::Receiver<FakeProducedBlock>) {
+        let (block_sender, receiver) = tokio::sync::mpsc::channel(100);
+        (Self { block_sender }, receiver)
+    }
+}
+
+#[async_trait]
+impl BlockProducer for FakeBlockProducer {
+    async fn produce_and_execute_block(
+        &self,
+        height: BlockHeight,
+        block_time: Tai64,
+        _source: TransactionsSource,
+    ) -> anyhow::Result<UncommittedResult<Changes>> {
+        self.block_sender
+            .send(FakeProducedBlock::New(height, block_time))
+            .await
+            .unwrap();
+        Ok(UncommittedResult::new(
+            ExecutionResult {
+                block: Default::default(),
+                skipped_transactions: Default::default(),
+                tx_status: Default::default(),
+                events: Default::default(),
+            },
+            Default::default(),
+        ))
+    }
+
+    async fn produce_predefined_block(
+        &self,
+        block: &Block,
+    ) -> anyhow::Result<UncommittedResult<Changes>> {
+        self.block_sender
+            .send(FakeProducedBlock::Predefined(block.clone()))
+            .await
+            .unwrap();
+        Ok(UncommittedResult::new(
+            ExecutionResult {
+                block: block.clone(),
+                skipped_transactions: Default::default(),
+                tx_status: Default::default(),
+                events: Default::default(),
+            },
+            Default::default(),
+        ))
+    }
+}
+
+fn block_for_height(height: u32) -> Block {
+    let mut header = PartialBlockHeader::default();
+    header.consensus.height = height.into();
+    let transactions = vec![];
+    Block::new(header, transactions, Default::default(), Default::default()).unwrap()
+}
+
+#[tokio::test]
+async fn consensus_service__run__will_include_predefined_blocks_before_new_blocks() {
+    // given
+    let blocks: [(BlockHeight, Block); 3] = [
+        (2u32.into(), block_for_height(2)),
+        (3u32.into(), block_for_height(3)),
+        (4u32.into(), block_for_height(4)),
```

Comment:
Could we also test the case when we produce some blocks first form the interval, then we have predefined blocks, please?